src/SwedbankPay.Sdk/SwedbankPayHttpClient.cs

```diff
@@ -0,0 +1,227 @@
+﻿using System;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+
+using SwedbankPay.Sdk.Exceptions;
+
+namespace SwedbankPay.Sdk
+{
+    internal class SwedbankPayHttpClient
+    {
+        private readonly HttpClient client;
+        private readonly ILogger logger;
+
+
+        internal SwedbankPayHttpClient(HttpClient client, ILogger logger)
+        {
+            this.client = client;
+            this.logger = logger;
+        }
+
+
+        internal async Task<string> GetRaw(string url)
+        {
+            var msg = new HttpRequestMessage(HttpMethod.Get, url);
+            msg.Headers.Add("Accept", "application/json");
+
+            HttpResponseMessage response;
+            try
+            {
+                response = await this.client.SendAsync(msg);
+            }
+            catch (HttpRequestException e)
+            {
+                throw new BadRequestException(e);
+            }
+            catch (TaskCanceledException te)
+            {
+                throw new ApiTimeOutException(te);
+            }
+
+            if (response.IsSuccessStatusCode)
+            {
+                var res = await response.Content.ReadAsStringAsync();
+                var res2 = JToken.Parse(res).ToString(Formatting.Indented);
+                this.logger.LogInformation(res2);
+                return res2;
+            }
+
+            return null;
+        }
+
+
+        internal async Task<TResponse> HttpGet<TResponse>(string url, Func<ProblemsContainer, Exception> onError)
+            where TResponse : new()
+        {
+            return await HttpRequest<TResponse>(HttpMethod.Get, url, onError);
+        }
+
+
+        internal async Task<TResponse> HttpPatch
+            <TPayLoad, TResponse>(string url, Func<ProblemsContainer, Exception> onError, TPayLoad payload)
+            where TResponse : new()
+        {
+            return await HttpRequest<TResponse>(new HttpMethod("PATCH"), url, onError, payload);
+        }
+
+
+        internal async Task<TResponse> HttpPost
+            <TPayLoad, TResponse>(string url, Func<ProblemsContainer, Exception> onError, TPayLoad payload)
+            where TResponse : new()
+        {
+            return await HttpRequest<TResponse>(HttpMethod.Post, url, onError, payload);
+        }
+
+
+        internal async Task<T> HttpRequest
+            <T>(string httpMethod, string url, Func<ProblemsContainer, Exception> onError, object payload = null)
+            where T : new()
+        {
+            return await HttpRequest<T>(new HttpMethod(httpMethod), url, onError, payload);
+        }
+
+
+        internal async Task<T> HttpRequest
+            <T>(HttpMethod httpMethod, string url, Func<ProblemsContainer, Exception> onError, object payload = null)
```

Comment:
What advantages does the onError() approach give us? Why not just let the caller handle this themselves by catching whatever we throw at them?

Reply:
OnError throws the exception that the SDK tells it to throw. For instance when a request is successful but the api returns a "Problem". E.g.` PaymentOrder.Create` throws a `CouldNotPlacePaymentOrderException`

It's still up to the caller to handle this exception themselves.

Reply:
That still doesn't explain why using the "OnError approach" is advantageous to simply throwing. As a side note OnError should be renamed to something like ExceptionToBeThrownOnError, perhaps? Is the lambda meant to actually do something in addition to returning an exception, or just return an exception?